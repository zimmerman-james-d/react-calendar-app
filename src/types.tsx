export interface EventDefinition {
  id: string;
  groupId?: string;
  title: string;
  deleted?: boolean;

  // --- Date Definition ---
  // For single, specific date events
  date?: string; 

  // For recurring events with specific weekly patterns
  recurrence?: {
    startRecur: string;
    endRecur: string;
    weeklySelections: number[][];
    recurrenceCycle: number;
  };

  // For events relative to another single event
  relativeTo?: {
    targetId: string; // ID of the target EventDefinition or "start-date"
    offset: number;   // Can be positive or negative
  };
  
  // For recurring events relative to another recurring series OR single event
  relativeRecurrence?: {
    targetId?: string; // ID of the target EventDefinition (single event or "start-date")
    targetGroupId?: string; // groupId of the target EventDefinition series
    targetType: 'single' | 'group'; // New field to differentiate
    daysBefore: boolean;
    daysAfter: boolean;
    dayOf: boolean;
    beforeOffset: number;
    afterOffset: number;
  };

  // --- Exceptions ---
  // Stores modifications to specific instances of a recurring event
  exceptions?: Record<string, string>; 
}
