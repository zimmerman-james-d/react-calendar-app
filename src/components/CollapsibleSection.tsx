import React, { useState, ReactNode } from 'react';

interface CollapsibleSectionProps {
  title: string;
  children: ReactNode;
  initialOpen?: boolean;
}

export function CollapsibleSection({ title, children, initialOpen = false }: CollapsibleSectionProps) {
  const [isOpen, setIsOpen] = useState(initialOpen);

  return (
    <div className="collapsible-section">
      <div className="collapsible-header" onClick={() => setIsOpen(!isOpen)}>
        <h2>{title}</h2>
        <span className="collapsible-icon">{isOpen ? 'âˆ’' : '+'}</span>
      </div>
      <div className={`collapsible-content ${isOpen ? 'open' : ''}`}>
        {children}
      </div>
    </div>
  );
}
