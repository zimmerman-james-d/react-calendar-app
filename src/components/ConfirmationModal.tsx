import React from 'react';

interface ConfirmationModalProps {
  isOpen: boolean;
  message: string;
  onConfirm: () => void;
  onCancel: () => void;
}

export function ConfirmationModal({ isOpen, message, onConfirm, onCancel }: ConfirmationModalProps) {
  if (!isOpen) return null;

  return (
    <div className="modal-overlay">
      <div className="modal-content">
        <h4>Confirmation</h4>
        <p>{message}</p>
        <div className="modal-actions">
          <button className="modal-button submit" onClick={onConfirm}>Delete</button>
          <button className="modal-button cancel" onClick={onCancel}>Cancel</button>
        </div>
      </div>
    </div>
  );
}
