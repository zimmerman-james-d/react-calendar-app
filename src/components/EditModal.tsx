import React, { useState } from 'react';
import { EventDefinition } from '../types';
import { EventInput } from '@fullcalendar/core';
import { daysOfWeek } from './RecurringEventForm';

interface EditModalProps {
    isOpen: boolean;
    event: EventDefinition | undefined;
    events: EventInput[];
    eventDefinitions: EventDefinition[];
    dateType: string;
    startDate: string;
    onConfirm: () => void;
    onCancel: () => void;
}

export function EditModal({ isOpen, event, events, eventDefinitions, startDate, dateType, onConfirm, onCancel }: EditModalProps) {
    if (!isOpen) return null;
    const [title, setTitle] = useState(event?.title);
    const [specificDate, setSpecificDate] = useState(event?.date);

    const [relativeOffset, setRelativeOffset] = useState<number>(event?.relativeTo?.offset || 1);
    const [relativeDirection, setRelativeDirection] = useState<'after' | 'before' | 'same'>('after');
    const [relativeTargetEventId, setRelativeTargetEventId] = useState<string>(event?.relativeTo?.targetId || '');
    const [startRecur, setStartRecur] = useState(event?.recurrence?.startRecur);
    const [endRecur, setEndRecur] = useState(event?.recurrence?.endRecur);
    const [recurrenceCycle, setRecurrenceCycle] = useState(event?.recurrence?.weeklySelections.length || 2);
    const [weeklySelections, setWeeklySelections] = useState<number[][]>(event?.recurrence?.weeklySelections || [[], []]);

    const [isDaysBeforeEnabled, setIsDaysBeforeEnabled] = useState(event?.relativeRecurrence?.daysBefore);
    const [daysBefore, setDaysBefore] = useState(event?.relativeRecurrence?.beforeOffset);
    const [isDaysAfterEnabled, setIsDaysAfterEnabled] = useState(event?.relativeRecurrence?.daysAfter);
    const [daysAfter, setDaysAfter] = useState(event?.relativeRecurrence?.afterOffset);
    const [isDayOfEnabled, setIsDayOfEnabled] = useState(event?.relativeRecurrence?.daysAfter);
    const [relativeTargetGroupId, setRelativeTargetGroupId] = useState<string>(event?.relativeRecurrence?.targetGroupId || '');
    const [relativeTargetId, setRelativeTargetId] = useState<string>(event?.relativeRecurrence?.targetId || '');

    const handleDayToggle = (dayValue: number, weekIndex: number) => {
        setWeeklySelections(currentSelections => {
            const newSelections = [...currentSelections];
            const week = [...(newSelections[weekIndex] || [])];
            if (week.includes(dayValue)) {
                newSelections[weekIndex] = week.filter(d => d !== dayValue);
            } else {
                newSelections[weekIndex] = [...week, dayValue];
            }
            return newSelections;
        });
    };


    return (
        <div className="modal-overlay">
            <div className="modal-content">
                <h4>{event?.title}</h4>
                <div className="form-group">
                    <label htmlFor="event-title">Event Name</label>
                    <input type="text" id="event-title" value={title} onChange={(e) => setTitle(e.target.value)} />
                </div>
                {dateType === 'specific' && (
                    <div className="form-group">
                        <label htmlFor="event-date">Date</label>
                        <input type="date" id="event-date" value={specificDate} onChange={(e) => setSpecificDate(e.target.value)} />
                    </div>
                )}
                {dateType === 'relative' && (
                    <div className="form-group">
                        <label>Relative Date</label>
                        <div className="relative-date-controls">
                            <input
                                type="number"
                                className="relative-offset-input"
                                value={relativeOffset}
                                onChange={(e) => setRelativeOffset(Number(e.target.value))}
                                disabled={relativeDirection === 'same'}
                            />
                            <select className="relative-direction-select" value={relativeDirection} onChange={(e) => setRelativeDirection(e.target.value as any)}>
                                <option value="after">Days After</option>
                                <option value="before">Days Before</option>
                                <option value="same">On the same day as</option>
                            </select>
                        </div>
                        <select className="relative-target-select" value={relativeTargetEventId} onChange={(e) => setRelativeTargetEventId(e.target.value)}>
                            <option value="">Select an event...</option>
                            {startDate && (
                                <option value="start-date">
                                    Start Date ({startDate})
                                </option>
                            )}
                            {events.map((event) => (
                                <option key={event.id} value={event.id}>
                                    {event.title} ({event.date?.toString()})
                                </option>
                            ))}
                        </select>
                    </div>
                )}
                {dateType === 'specific-recurrance' && (
                    <>
                        <div className="form-group">
                            <label htmlFor="recurrence-cycle">Repeats Every</label>
                            <select id="recurrence-cycle" value={recurrenceCycle} onChange={(e) => setRecurrenceCycle(Number(e.target.value))}>
                                <option value={1}>1 Week</option>
                                <option value={2}>2 Weeks</option>
                                <option value={3}>3 Weeks</option>
                                <option value={4}>4 Weeks</option>
                            </select>
                        </div>
                        {Array.from({ length: recurrenceCycle }).map((_, index) => (
                            <div className="form-group" key={index}>
                                <label>Week {index + 1} Repeats On</label>
                                <div className="day-picker">
                                    {daysOfWeek.map(day => (
                                        <button
                                            key={day.value}
                                            className={`day-button ${weeklySelections[index]?.includes(day.value) ? 'selected' : ''}`}
                                            onClick={() => handleDayToggle(day.value, index)}>
                                            {day.label}
                                        </button>
                                    ))}
                                </div>
                            </div>
                        ))}
                        <div className="form-group">
                            <label htmlFor="start-recur">Start Date</label>
                            <input type="date" id="start-recur" value={startRecur} onChange={(e) => setStartRecur(e.target.value)} />
                        </div>
                        <div className="form-group">
                            <label htmlFor="end-recur">End Date</label>
                            <input type="date" id="end-recur" value={endRecur} onChange={(e) => setEndRecur(e.target.value)} />
                        </div>
                    </>
                )}
                {(dateType === 'relative-group' || dateType === 'relative-single') && (
                    <>
                        <div className="form-group">
                            <label htmlFor={dateType === 'relative-group' ? 'relative-target-group' : 'relative-target-single'}>Relative To</label>
                            {dateType === 'relative-group' && (
                                <select id="relative-target-group" className="relative-target-select" value={relativeTargetGroupId} onChange={(e) => setRelativeTargetGroupId(e.target.value)}>
                                    <option value="">Select a recurring event...</option>
                                    {eventDefinitions.filter(def => def.recurrence).map((def) => (
                                        <option key={def.id} value={def.groupId}>
                                            {def.title}
                                        </option>
                                    ))}
                                </select>
                            )}
                            {dateType === 'relative-single' && (
                                <select id="relative-target-single" className="relative-target-select" value={relativeTargetId} onChange={(e) => setRelativeTargetId(e.target.value)}>
                                    <option value="">Select a single event or start date...</option>
                                    <option value="start-date">Start Date</option>
                                    {eventDefinitions.filter(def => def.date).map((def) => (
                                        <option key={def.id} value={def.id}>
                                            {def.title} ({def.date})
                                        </option>
                                    ))}
                                </select>
                            )}
                        </div>
                        <div className="relative-option-group">
                            <input type="checkbox" id="days-before-check" checked={isDaysBeforeEnabled} onChange={() => setIsDaysBeforeEnabled(!isDaysBeforeEnabled)} />
                            <input type="number" value={daysBefore} onChange={(e) => setDaysBefore(Number(e.target.value))} disabled={!isDaysBeforeEnabled} />
                            <label htmlFor="days-before-check">Days Before</label>
                        </div>
                        <div className="relative-option-group">
                            <input type="checkbox" id="days-after-check" checked={isDaysAfterEnabled} onChange={() => setIsDaysAfterEnabled(!isDaysAfterEnabled)} />
                            <input type="number" value={daysAfter} onChange={(e) => setDaysAfter(Number(e.target.value))} disabled={!isDaysAfterEnabled} />
                            <label htmlFor="days-after-check">Days After</label>
                        </div>
                        <div className="relative-option-group">
                            <input type="checkbox" id="day-of-check" checked={isDayOfEnabled} onChange={() => setIsDayOfEnabled(!isDayOfEnabled)} />
                            <label htmlFor="day-of-check">Day Of</label>
                        </div>
                    </>
                )}
                <div className="modal-actions">
                    <button className="modal-button submit" onClick={onConfirm}>Update</button>
                    <button className="modal-button cancel" onClick={onCancel}>Cancel</button>
                </div>
            </div>
        </div>
    );
}
