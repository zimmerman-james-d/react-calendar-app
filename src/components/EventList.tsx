import React from 'react';
import { EventDefinition } from '../types';

interface EventListProps {
  eventDefinitions: EventDefinition[];
  onRemoveEventDefinition: (id: string) => void;
  onRestoreEventDefinition: (id: string) => void;
  onPermanentDeleteEventDefinition: (id: string) => void; // New prop
}

const dayMap: { [key: number]: string } = {
  0: 'Sun', 1: 'Mon', 2: 'Tue', 3: 'Wed', 4: 'Thu', 5: 'Fri', 6: 'Sat'
};

// Helper function to format the recurrence rule into a readable string
export function formatRecurrenceRule(definition: EventDefinition, allDefinitions: EventDefinition[]): React.ReactNode {
  if (definition.recurrence) {
    const { startRecur, endRecur, weeklySelections, recurrenceCycle } = definition.recurrence;

    return (
      <div>
        <div style={{ display: 'flex', justifyContent: 'space-between', alignItems: 'center' }}>
          <span>{startRecur} - {endRecur}</span>
          <span style={{ textAlign: 'right' }}>{recurrenceCycle} Week{recurrenceCycle > 1 ? 's' : ''}</span>
        </div>
        {weeklySelections.map((days, index) => {
          const dayNames = days.map(d => dayMap[d]).join(', ');
          return <div key={index}>Week {index + 1}: {dayNames}</div>;
        })}
      </div>
    );
  }
  if (definition.relativeRecurrence) {
    let targetTitle = "";
    if (definition.relativeRecurrence.targetType === 'group') {
        const targetDef = allDefinitions.find(d => d.groupId === definition.relativeRecurrence?.targetGroupId);
        targetTitle = targetDef ? `"${targetDef.title}"` : "another series";
    } else if (definition.relativeRecurrence.targetType === 'single') {
        if (definition.relativeRecurrence.targetId === 'start-date') {
            targetTitle = "the Start Date";
        } else {
            const targetDef = allDefinitions.find(d => d.id === definition.relativeRecurrence?.targetId);
            targetTitle = targetDef ? `"${targetDef.title}"` : "another event";
        }
    }
    
    const parts: string[] = [];
    if (definition.relativeRecurrence.daysBefore) {
        parts.push(`${definition.relativeRecurrence.beforeOffset} day(s) before`);
    }
    if (definition.relativeRecurrence.dayOf) {
        parts.push('the day of');
    }
    if (definition.relativeRecurrence.daysAfter) {
        parts.push(`${definition.relativeRecurrence.afterOffset} day(s) after`);
    }

    let detailString = '';
    if (parts.length === 1) {
        detailString = parts[0];
    } else if (parts.length === 2) {
        detailString = parts.join(' and ');
    } else if (parts.length > 2) {
        detailString = `${parts.slice(0, -1).join(', ')}, and ${parts[parts.length - 1]}`;
    }

    return `Repeats ${detailString} ${targetTitle}`;
  }
  if (definition.relativeTo) {
    const { targetId, offset } = definition.relativeTo;
    let targetTitle = "another event";
    if (targetId === 'start-date') {
      targetTitle = "the Start Date";
    }
    else {
      const targetDef = allDefinitions.find(d => d.id === targetId);
      if (targetDef) {
        targetTitle = `"${targetDef.title}"`;
      }
    }

    if (offset === 0) {
      return `On the same day as ${targetTitle}`;
    }
    const direction = offset > 0 ? 'after' : 'before';
    return `${Math.abs(offset)} day(s) ${direction} ${targetTitle}`;
  }
  return 'Event rule not specified';
}

export function EventList({ eventDefinitions, onRemoveEventDefinition, onRestoreEventDefinition }: EventListProps) {
  const activeEvents = eventDefinitions.filter(def => !def.deleted);
  const deletedEvents = eventDefinitions.filter(def => def.deleted);

  return (
    <>
      <ul className="event-list">
        {activeEvents.map((def) => (
          <li key={def.id} className="event-list-item">
            <strong>{def.title}</strong>
            <br />
            <small>
              {def.date ? def.date : formatRecurrenceRule(def, eventDefinitions)}
            </small>
            <button className="remove-event-button" onClick={() => onRemoveEventDefinition(def.id)}>Remove</button>
          </li>
        ))}
      </ul>

      {deletedEvents.length > 0 && (
        <div className="deleted-events-section">
          <h3>Deleted Events</h3>
          <ul className="event-list">
            {deletedEvents.map((def) => (
              <li key={def.id} className={`event-list-item ${def.deleted ? 'deleted-event' : ''}`} style={{ position: 'relative' }}>
                <strong>{def.title}</strong>
                <br />
                <small>
                  {def.date ? def.date : formatRecurrenceRule(def, eventDefinitions)}
                </small>
                <button className="restore-event-button" onClick={() => onRestoreEventDefinition(def.id)} style={{ position: 'absolute', top: '5px', right: '5px' }}>Restore</button>
              </li>
            ))}
          </ul>
        </div>
      )}
    </>
  );
}
