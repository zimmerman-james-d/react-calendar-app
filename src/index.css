@tailwind base;
@tailwind components;
@tailwind utilities;

.app-container {
  display: flex;
  height: 98vh;
  width: 98vw;
  overflow: hidden;
}

.main-content {
  flex-grow: 1;
  padding: 20px;
  display: flex;
  flex-direction: column;
}

.calendar-container {
  flex-grow: 1;
}

.sidebar-container {
  position: relative;
  display: flex;
  height: 100%;
}

.sidebar {
  background-color: #f5f5f5;
  padding: 0 20px 0 20px;
  transition: width 0.3s ease;
  width: 300px;
  height: 100%;
  white-space: nowrap;
  display: flex;
  flex-direction: column;
  overflow: hidden;
}

.sidebar-header {
    padding-top: 5px;
    flex-shrink: 0;
}

.sidebar-content {
    flex-grow: 1;
    overflow-y: auto;
}

.sidebar-footer {
    padding-bottom: 30px;
    flex-shrink: 0;
}


.sidebar-container.closed .sidebar {
  width: 0;
}

.sidebar-toggle-button {
  position: absolute;
  top: 50%;
  left: 100%; 
  transform: translateX(-50%);
  z-index: 10;
  height: 40px;
  width: 40px;
  border-radius: 50%;
  background-color: #2c3e50;
  color: white;
  border: none;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: center;
  font-size: 24px;
  line-height: 1;
  padding: 0;
  transition: transform 0.3s ease;
}

.sidebar-container.closed .sidebar-toggle-button {
  transform: translateX(-20px);
}

.top-level-form-group {
  margin-bottom: 20px;
  padding-bottom: 20px;
  border-bottom: 2px solid #ddd;
}

.sidebar-header .top-level-form-group:last-child {
    margin-bottom: 0;
}


.form-group {
  margin-bottom: 15px;
}

.form-group label,
.top-level-form-group label {
  display: block;
  font-weight: bold;
  margin-bottom: 5px;
  color: #333;
}

.form-group input,
.form-group select,
.top-level-form-group input {
  width: 100%;
  padding: 8px;
  border: 1px solid #ccc;
  border-radius: 4px;
  box-sizing: border-box;
}

.save-event-button {
  width: 100%;
  padding: 10px;
  background-color: #2c3e50;
  color: white;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-size: 16px;
}

.save-event-button:hover {
  background-color: #3a506b;
}

.collapsible-section {
  border-bottom: 1px solid #ddd;
  padding-bottom: 10px;
  margin-bottom: 10px;
}

.collapsible-header {
  display: flex;
  justify-content: space-between;
  align-items: center;
  cursor: pointer;
}

.collapsible-header h2 {
  margin: 0;
}

.collapsible-icon {
  font-size: 24px;
  font-weight: bold;
  color: #555;
}

.collapsible-content {
  max-height: 0;
  overflow: hidden;
  transition: max-height 0.3s ease-out;
  padding-top: 0;
}

.collapsible-content.open {
  max-height: 900px;
  padding-top: 15px;
}

.day-picker {
  display: flex;
  justify-content: space-between;
  gap: 5px;
  transition: opacity 0.3s ease;
}

.day-picker.disabled {
  opacity: 0.5;
  pointer-events: none;
}

.day-button {
  flex-grow: 1;
  border: 1px solid #ccc;
  background-color: #fff;
  border-radius: 4px;
  padding: 8px 4px;
  cursor: pointer;
  transition: background-color 0.2s, color 0.2s;
}

.day-button.selected {
  background-color: #2c3e50;
  color: white;
  border-color: #2c3e50;
}

.week-toggle {
  display: flex;
  align-items: center;
  margin-bottom: 8px;
}

.week-toggle input[type="checkbox"] {
  margin-right: 8px;
  width: 16px;
  height: 16px;
  cursor: pointer;
}

.week-toggle label {
  font-weight: bold;
  color: #333;
  margin-bottom: 0;
  cursor: pointer;
}

.event-list {
  list-style: none;
  padding: 0;
  margin: 0;
}

.event-list-item {
  background-color: #fff;
  border: 1px solid #eee;
  padding: 10px;
  border-radius: 4px;
  margin-bottom: 8px;
  white-space: normal;
  position: relative;
}

.event-action-button-base {
  background: none;
  border: none;
  font-size: 0.8em;
  cursor: pointer;
  padding: 2px 5px;
  border-radius: 3px;
}

.remove-event-button {
  position: absolute;
  bottom: 5px;
  right: 5px;
  color: red;
}

.remove-event-button:hover {
  background-color: rgba(255, 0, 0, 0.1);
}


.event-list-item strong {
  color: #333;
}

.event-list-item small {
  color: #777;
}

.relative-date-controls {
  display: flex;
  gap: 10px;
  margin-bottom: 10px;
}

.relative-offset-input {
  width: 80px;
}

.relative-direction-select {
  flex-grow: 1;
}

.relative-target-select {
  width: 100%;
}

.relative-option-group {
    display: flex;
    align-items: center;
    gap: 10px;
    margin-bottom: 10px;
}

.relative-option-group input[type="number"] {
    width: 70px;
}

.relative-option-group label {
    margin-bottom: 0;
    font-weight: normal;
}

.save-load-controls {
    padding-top: 20px;
    border-top: 2px solid #ddd;
    display: flex;
    gap: 10px;
    flex-shrink: 0;
}

.save-button, .load-button {
    flex-grow: 1;
    padding: 10px;
    border: none;
    border-radius: 4px;
    cursor: pointer;
    font-size: 16px;
    color: white;
}

.save-button {
    background-color: #27ae60;
}
.save-button:hover {
    background-color: #229954;
}

.load-button {
    background-color: #3498db;
}
.load-button:hover {
    background-color: #2980b9;
}

.fc-event-main {
  background-color: #bee3f8 !important;
  border-color: #90cdf4 !important;
}

.fc-event-title {
  color: #000000 !important;
}

.modal-overlay {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: rgba(0, 0, 0, 0.5);
  display: flex;
  align-items: center;
  justify-content: center;
  z-index: 1000;
}

.modal-content {
  background-color: white;
  padding: 20px;
  border-radius: 8px;
  width: 90%;
  max-width: 400px;
  box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
}

.modal-content h4 {
  margin-top: 0;
  margin-bottom: 15px;
}

.password-input {
  width: 100%;
  padding: 10px;
  border: 1px solid #ccc;
  border-radius: 4px;
  margin-bottom: 15px;
  box-sizing: border-box;
}

.modal-actions {
  display: flex;
  justify-content: flex-end;
  gap: 10px;
}

.modal-button {
  padding: 8px 16px;
  border: none;
  border-radius: 4px;
  cursor: pointer;
}

.modal-button.submit {
  background-color: #2c3e50;
  color: white;
}

.modal-button.cancel {
  background-color: #e0e0e0;
}

.deleted-event {
  text-decoration: line-through;
  opacity: 0.6;
}

.deleted-events-section {
  margin-top: 20px;
  padding-top: 10px;
  border-top: 1px solid #eee;
}

.deleted-events-section h3 {
  color: #888;
  margin-bottom: 10px;
}

.restore-event-button {
  position: absolute;
  top: 5px;
  right: 5px;
  color: green;
}

.restore-event-button:hover {
  background-color: rgba(0, 128, 0, 0.1);
}

.delete-permanently-button {
  position: absolute;
  bottom: 5px;
  right: 5px;
  color: red;
}

.delete-permanently-button:hover {
  background-color: rgba(255, 0, 0, 0.1);
}
