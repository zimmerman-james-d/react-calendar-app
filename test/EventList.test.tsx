import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { EventList, formatRecurrenceRule } from '../src/components/EventList';
import { EventDefinition } from '../src/types';

const dayMap: { [key: number]: string } = {
  0: 'Sun', 1: 'Mon', 2: 'Tue', 3: 'Wed', 4: 'Thu', 5: 'Fri', 6: 'Sat'
};

describe('EventList Component', () => {
  const mockOnDeleteEventDefinition = jest.fn();
  const mockOnRestoreEventDefinition = jest.fn();

  const mockEventDefinitions: EventDefinition[] = [
    { id: '1', title: 'Test Event 1', date: '2025-10-20', deleted: false },
    { id: '2', title: 'Test Event 2', date: '2025-10-21', deleted: true },
  ];

  beforeEach(() => {
    mockOnDeleteEventDefinition.mockClear();
    mockOnRestoreEventDefinition.mockClear();
  });

  it('should render active event definitions and call onDeleteEventDefinition when delete button is clicked', () => {
    render(
      <EventList 
        eventDefinitions={mockEventDefinitions} 
        onDeleteEventDefinition={mockOnDeleteEventDefinition} 
        onRestoreEventDefinition={mockOnRestoreEventDefinition}
      />
    );

    const listItems = screen.getAllByRole('listitem');
    const activeListItems = listItems.filter(item => !item.classList.contains('deleted-event'));
    const deletedListItems = listItems.filter(item => item.classList.contains('deleted-event'));

    expect(activeListItems.length).toBe(1);
    expect(activeListItems[0]).toHaveTextContent('Test Event 1');

    expect(deletedListItems.length).toBe(1);
    expect(deletedListItems[0]).toHaveTextContent('Test Event 2');
    expect(deletedListItems[0]).toHaveClass('deleted-event');

    // Find the delete button for Test Event 1 and click it
    const deleteButtons = activeListItems[0].querySelector('.delete-event-button');
    if (deleteButtons) {
      fireEvent.click(deleteButtons);
    }

    // Assert that onDeleteEventDefinition was called with the correct ID
    expect(mockOnDeleteEventDefinition).toHaveBeenCalledTimes(1);
    expect(mockOnDeleteEventDefinition).toHaveBeenCalledWith('1');
  });

  it('should render deleted event definitions and call onRestoreEventDefinition when restore button is clicked', () => {
    render(
      <EventList 
        eventDefinitions={mockEventDefinitions} 
        onDeleteEventDefinition={mockOnDeleteEventDefinition} 
        onRestoreEventDefinition={mockOnRestoreEventDefinition}
      />
    );

    // Check if deleted events section and event are rendered
    expect(screen.getByText('Deleted Events')).toBeInTheDocument();
    const deletedEventTextElement = screen.getByText('Test Event 2');
    const deletedEventListItem = deletedEventTextElement.closest('li');
    expect(deletedEventListItem).toBeInTheDocument();
    expect(deletedEventListItem).toHaveClass('deleted-event');

    // Find the restore button for Test Event 2 and click it
    const restoreButton = screen.getByText('Restore');
    fireEvent.click(restoreButton);

    // Assert that onRestoreEventDefinition was called with the correct ID
    expect(mockOnRestoreEventDefinition).toHaveBeenCalledTimes(1);
    expect(mockOnRestoreEventDefinition).toHaveBeenCalledWith('2');
  });
});

describe('formatRecurrenceRule', () => {
  const mockAllDefinitions: EventDefinition[] = [
    { id: 'base-single', title: 'Base Single', date: '2025-01-01' },
    { id: 'base-group', groupId: 'group-abc', title: 'Base Group', recurrence: { startRecur: '2025-01-01', endRecur: '2025-01-07', weeklySelections: [[1]], recurrenceCycle: 1 } },
  ];

  it('should format a recurrence rule correctly', () => {
    const definition: EventDefinition = {
      id: 'rec-1',
      title: 'Weekly Meeting',
      recurrence: {
        startRecur: '2025-10-06',
        endRecur: '2025-10-20',
        weeklySelections: [[1, 3]], // Mondays and Wednesdays
        recurrenceCycle: 1,
      },
    };
    const result = formatRecurrenceRule(definition, []);
    expect(result).toMatchSnapshot();
  });

  it('should format a recurrence rule with recurrenceCycle > 1 correctly', () => {
    const definition: EventDefinition = {
      id: 'rec-2',
      title: 'Bi-weekly Meeting',
      recurrence: {
        startRecur: '2025-10-06',
        endRecur: '2025-10-20',
        weeklySelections: [[1], [3]], // Week 1: Monday, Week 2: Wednesday
        recurrenceCycle: 2,
      },
    };
    const result = formatRecurrenceRule(definition, []);
    expect(result).toMatchSnapshot();
  });

  it('should format a relative recurrence to a group with daysBefore', () => {
    const definition: EventDefinition = {
      id: 'rel-rec-1',
      title: 'Prep Meeting',
      relativeRecurrence: {
        targetGroupId: 'group-abc',
        targetType: 'group',
        daysBefore: true,
        beforeOffset: 2,
        daysAfter: false,
        dayOf: false,
        afterOffset: 0,
      },
    };
    const result = formatRecurrenceRule(definition, mockAllDefinitions);
    expect(result).toBe('Repeats 2 day(s) before "Base Group"');
  });

  it('should format a relative recurrence to a group with dayOf', () => {
    const definition: EventDefinition = {
      id: 'rel-rec-2',
      title: 'Main Event',
      relativeRecurrence: {
        targetGroupId: 'group-abc',
        targetType: 'group',
        dayOf: true,
        daysBefore: false,
        daysAfter: false,
        beforeOffset: 0,
        afterOffset: 0,
      },
    };
    const result = formatRecurrenceRule(definition, mockAllDefinitions);
    expect(result).toBe('Repeats the day of "Base Group"');
  });

  it('should format a relative recurrence to a group with daysAfter', () => {
    const definition: EventDefinition = {
      id: 'rel-rec-3',
      title: 'Follow-up',
      relativeRecurrence: {
        targetGroupId: 'group-abc',
        targetType: 'group',
        daysAfter: true,
        afterOffset: 3,
        daysBefore: false,
        dayOf: false,
        beforeOffset: 0,
      },
    };
    const result = formatRecurrenceRule(definition, mockAllDefinitions);
    expect(result).toBe('Repeats 3 day(s) after "Base Group"');
  });

  it('should format a relative recurrence to a single event (start-date)', () => {
    const definition: EventDefinition = {
      id: 'rel-rec-4',
      title: 'Project Start',
      relativeRecurrence: {
        targetId: 'start-date',
        targetType: 'single',
        dayOf: true,
        daysBefore: false,
        daysAfter: false,
        beforeOffset: 0,
        afterOffset: 0,
      },
    };
    const result = formatRecurrenceRule(definition, mockAllDefinitions);
    expect(result).toBe('Repeats the day of the Start Date');
  });

  it('should format a relative recurrence to a single event (specific ID)', () => {
    const definition: EventDefinition = {
      id: 'rel-rec-5',
      title: 'Review',
      relativeRecurrence: {
        targetId: 'base-single',
        targetType: 'single',
        daysAfter: true,
        afterOffset: 1,
        daysBefore: false,
        beforeOffset: 0,
        dayOf: false,
      },
    };
    const result = formatRecurrenceRule(definition, mockAllDefinitions);
    expect(result).toBe('Repeats 1 day(s) after "Base Single"');
  });

  it('should format a relative recurrence with multiple parts', () => {
    const definition: EventDefinition = {
      id: 'rel-rec-6',
      title: 'Complex Relative',
      relativeRecurrence: {
        targetGroupId: 'group-abc',
        targetType: 'group',
        daysBefore: true,
        beforeOffset: 1,
        dayOf: true,
        daysAfter: true,
        afterOffset: 1,
      },
    };
    const result = formatRecurrenceRule(definition, mockAllDefinitions);
    expect(result).toBe('Repeats 1 day(s) before, the day of, and 1 day(s) after "Base Group"');
  });

  it('should format a relativeTo event (start-date)', () => {
    const definition: EventDefinition = {
      id: 'rel-to-1',
      title: 'Initial Setup',
      relativeTo: { targetId: 'start-date', offset: 0 },
    };
    const result = formatRecurrenceRule(definition, mockAllDefinitions);
    expect(result).toBe('On the same day as the Start Date');
  });

  it('should format a relativeTo event (specific ID, positive offset)', () => {
    const definition: EventDefinition = {
      id: 'rel-to-2',
      title: 'Follow-up Task',
      relativeTo: { targetId: 'base-single', offset: 7 },
    };
    const result = formatRecurrenceRule(definition, mockAllDefinitions);
    expect(result).toBe('7 day(s) after "Base Single"');
  });

  it('should format a relativeTo event (specific ID, negative offset)', () => {
    const definition: EventDefinition = {
      id: 'rel-to-3',
      title: 'Pre-meeting Prep',
      relativeTo: { targetId: 'base-single', offset: -2 },
    };
    const result = formatRecurrenceRule(definition, mockAllDefinitions);
    expect(result).toBe('2 day(s) before "Base Single"');
  });

  it('should return "Event rule not specified" for unknown rule types', () => {
    const definition: EventDefinition = {
      id: 'unknown-1',
      title: 'Unknown Event',
    };
    const result = formatRecurrenceRule(definition, mockAllDefinitions);
    expect(result).toBe('Event rule not specified');
  });
});
