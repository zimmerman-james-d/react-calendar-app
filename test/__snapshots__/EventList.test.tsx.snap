// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatRecurrenceRule should format a recurrence rule correctly 1`] = `
<div>
  <div
    style={
      {
        "alignItems": "center",
        "display": "flex",
        "justifyContent": "space-between",
      }
    }
  >
    <span>
      2025-10-06
       - 
      2025-10-20
    </span>
    <span
      style={
        {
          "textAlign": "right",
        }
      }
    >
      1
       Week
      
    </span>
  </div>
  <div>
    Week 
    1
    : 
    Mon, Wed
  </div>
</div>
`;

exports[`formatRecurrenceRule should format a recurrence rule with recurrenceCycle > 1 correctly 1`] = `
<div>
  <div
    style={
      {
        "alignItems": "center",
        "display": "flex",
        "justifyContent": "space-between",
      }
    }
  >
    <span>
      2025-10-06
       - 
      2025-10-20
    </span>
    <span
      style={
        {
          "textAlign": "right",
        }
      }
    >
      2
       Week
      s
    </span>
  </div>
  <div>
    Week 
    1
    : 
    Mon
  </div>
  <div>
    Week 
    2
    : 
    Wed
  </div>
</div>
`;
